subscription.go

```diff
@@ -0,0 +1,219 @@
+		SubscriptionID:     s.SubscriptionID,
+		TimestampsToReturn: ts,
+		ItemsToCreate:      items,
+	}
+
+	var res *ua.CreateMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) Unmonitor(monitoredItemIDs ...uint32) (*ua.DeleteMonitoredItemsResponse, error) {
+	req := &ua.DeleteMonitoredItemsRequest{
+		MonitoredItemIDs: monitoredItemIDs,
+		SubscriptionID:   s.SubscriptionID,
+	}
+	var res *ua.DeleteMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func publishOnceForSubscription(s *Subscription, acks []*ua.SubscriptionAcknowledgement) (*ua.PublishResponse, error) {
+	if acks == nil {
+		acks = []*ua.SubscriptionAcknowledgement{}
+	}
+	req := &ua.PublishRequest{
+		SubscriptionAcknowledgements: acks,
+	}
+
+	var res *ua.PublishResponse
+	err := s.c.sendWithTimeout(req, s.getPublishTimeout(), func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) getPublishTimeout() time.Duration {
+	keepaliveIntervalMs := uint64(s.RevisedPublishingInterval) *
+		uint64(s.RevisedMaxKeepAliveCount)
+	if keepaliveIntervalMs > math.MaxUint32 {
+		keepaliveIntervalMs = math.MaxUint32
+	}
+	requestedTimeout := time.Duration(keepaliveIntervalMs) * time.Millisecond
+	if requestedTimeout < s.c.cfg.RequestTimeout {
+		return s.c.cfg.RequestTimeout
+	}
+	return requestedTimeout
+}
+
+// Run() starts an infinite loop that sends PublishRequests and delivers received
+// notifications to registered Subscriptions.
+func (s *Subscription) Run(ctx context.Context) {
+	var acks []*ua.SubscriptionAcknowledgement
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			res, err := publishOnceForSubscription(s, acks)
+			if err != nil {
+				// StatusBadSequenceNumberUnknown means at least one ack has been submitted repeatedly
+				// Ignore the error. Acks will be cleared below
+				if err == ua.StatusBadSequenceNumberUnknown ||
+					// Ignore StatusBadTimeout. No need to do anything except continue the loop
+					err == ua.StatusBadTimeout ||
+					// Ignore StatusBadNoSubscription as probably the cause is that all subscriptions
+					// are already deleted, but the publishing loop is still running and will be stopped shortly
+					err == ua.StatusBadNoSubscription {
+				} else {
+					s.c.notifySubscriptions(ctx, err)
+				}
+			}
+
+			if res != nil {
+				// Prepare SubscriptionAcknowledgement for next PublishRequest
+				acks = make([]*ua.SubscriptionAcknowledgement, 0)
+				if res.AvailableSequenceNumbers != nil {
+					for _, i := range res.AvailableSequenceNumbers {
+						ack := &ua.SubscriptionAcknowledgement{
+							SubscriptionID: res.SubscriptionID,
+							SequenceNumber: i,
+						}
+						acks = append(acks, ack)
+					}
+				}
+			}
+
+			if err == nil {
+				s.c.notifySubscription(ctx, res)
```

Comment:
Why is `notifySubscription` not a method of `Subscription`? 

Either we only notify our subscription, then it should be a method. Or we notify other subscriptions in which case there should be only one run loop in the client.

Or am I missing something?

Reply:
I think I've covered this [here](https://github.com/gopcua/opcua/pull/207#discussion_r290176234)